### ISSUE TITLE
`Generator.send` method inferred type is wrong

### ISSUE THREAD

**Issue by kuyugama** [astral-sh/ty#3148]
### Summary

The `Generator.send` method should have type defined as Callable[[S | None], Y].
The first call of the `Generator.send` method should be with `None` as noted in the python docs https://docs.python.org/3/reference/expressions.html#generator.send

<img width="2244" height="1444" alt="Image" src="https://github.com/user-attachments/assets/0aa4aca3-2bee-4cc4-bb64-9c9b431bfbb7" />

```
error[invalid-argument-type]: Argument to bound method `send` is incorrect
  --> tytest.py:10:14
   |
 9 | x = some()
10 | print(x.send(None))
   |              ^^^^ Expected `str`, found `None`
11 | try:
12 |     print(x.send("test"))
   |
info: Method defined here
    --> stdlib/typing.pyi:1421:9
     |
1420 |     @abstractmethod
1421 |     def send(self, value: _SendT_contra, /) -> _YieldT_co:
     |         ^^^^       -------------------- Parameter declared here
1422 |         """Send a value into the generator.
1423 |         Return next yielded value or raise StopIteration.
     |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```

### Version
ty 0.0.26

**Comment by MeGaGiGaGon** [comment]
Transcribed for easy copy/pasting:
https://play.ty.dev/94f8fc0b-2260-4d48-863d-4ece7c2fb6a7
```py
from collections.abc import Generator

def some() -> Generator[int, str, str]:
    x = yield 1
    return x

x = some()
print(x.send(None))
try:
    print(x.send("test"))
except StopIteration as x:
    print(x.value)
```

The standard library type stubs are maintained in [typeshed](https://github.com/python/typeshed), so that would be the place to go for making an issue about adding `| None` to the method.

From what I can see no other type checker special cases this (tried pyright, pyrefly, mypy, zuban, and pycroscope). 

While it may be obvious in this snippet that `send` has not been called before, that is not true generally. Imagine a function like this that takes in that generator:
```py
def foo(gen: Generator[int, str, str]):
    gen.send(None)  # Should this be allowed?
```
There would be no way to know if `gen` has already been `send`ed or not, so it's not generally safe to `send` a `None` to a generator, without more knowledge than a type checker would have.

I did search the issues on typeshed but did not see one for this specifically, so you could open an issue there about it. I'm skeptical about the value of this, but if that is something that would be accepted then it would be fixed for all type checkers, including ty.

**Comment by carljm** [comment]
Thanks for the report! This is a known friction point with typing of generators in Python (see e.g. https://github.com/python/mypy/issues/11023), and all type checkers behave the same as ty here.

The core problem is that the correct typing of `send` is state-dependent; `None` should be allowed for the first call, but for subsequent calls only the expected type should be allowed. And it's not feasible in all cases for a type checker to know which call is the first; this state information is not encoded in the type system.

Type checkers could just always allow `None` as argument to `send`, but this would be unsound if we still inferred the type of `yield 1` as `str` inside the generator.

The usual workarounds are either to use `next(gen)` instead of `gen.send(None)` for the initial "start the generator" call, or to explicitly type your send type as `str | None` instead of `str` (but then you have to handle the possibility of receiving a `None` at your yields inside the generator.) The latter is really the only sound option, as typeshed annotations will allow `next(gen)` (which does send `None`) even for generators not annotated to receive `None`.

Closing as I don't think there's anything actionable for ty here -- any improvement to this situation will need to be ecosystem-wide.
